module/zfs/spa.c

```diff
@@ -6357,6 +6381,117 @@ spa_vdev_detach(spa_t *spa, uint64_t guid, uint64_t pguid, int replace_done)
 	return (error);
 }
 
+static int
+spa_vdev_initialize_impl(spa_t *spa, uint64_t guid, uint64_t cmd_type,
+    list_t *vd_list)
+{
+	ASSERT(MUTEX_HELD(&spa_namespace_lock));
+
+	spa_config_enter(spa, SCL_CONFIG | SCL_STATE, FTAG, RW_READER);
+
+	/* Look up vdev and ensure it's a leaf. */
+	vdev_t *vd = spa_lookup_by_guid(spa, guid, B_FALSE);
+	if (vd == NULL || vd->vdev_detached) {
+		spa_config_exit(spa, SCL_CONFIG | SCL_STATE, FTAG);
+		return (SET_ERROR(ENODEV));
+	} else if (!vd->vdev_ops->vdev_op_leaf || !vdev_is_concrete(vd)) {
+		spa_config_exit(spa, SCL_CONFIG | SCL_STATE, FTAG);
+		return (SET_ERROR(EINVAL));
+	} else if (!vdev_writeable(vd)) {
+		spa_config_exit(spa, SCL_CONFIG | SCL_STATE, FTAG);
+		return (SET_ERROR(EROFS));
+	}
+	mutex_enter(&vd->vdev_initialize_lock);
+	spa_config_exit(spa, SCL_CONFIG | SCL_STATE, FTAG);
+
+	/*
+	 * When we activate an initialize action we check to see
+	 * if the vdev_initialize_thread is NULL. We do this instead
+	 * of using the vdev_initialize_state since there might be
+	 * a previous initialization process which has completed but
+	 * the thread is not exited.
+	 */
+	if (cmd_type == POOL_INITIALIZE_DO &&
+	    (vd->vdev_initialize_thread != NULL ||
+	    vd->vdev_top->vdev_removing)) {
+		mutex_exit(&vd->vdev_initialize_lock);
+		return (SET_ERROR(EBUSY));
+	} else if (cmd_type == POOL_INITIALIZE_CANCEL &&
+	    (vd->vdev_initialize_state != VDEV_INITIALIZE_ACTIVE &&
+	    vd->vdev_initialize_state != VDEV_INITIALIZE_SUSPENDED)) {
+		mutex_exit(&vd->vdev_initialize_lock);
+		return (SET_ERROR(ESRCH));
+	} else if (cmd_type == POOL_INITIALIZE_SUSPEND &&
+	    vd->vdev_initialize_state != VDEV_INITIALIZE_ACTIVE) {
+		mutex_exit(&vd->vdev_initialize_lock);
+		return (SET_ERROR(ESRCH));
+	}
+
+	switch (cmd_type) {
+	case POOL_INITIALIZE_DO:
+		vdev_initialize(vd);
+		break;
+	case POOL_INITIALIZE_CANCEL:
+		vdev_initialize_stop(vd, VDEV_INITIALIZE_CANCELED, vd_list);
+		break;
+	case POOL_INITIALIZE_SUSPEND:
+		vdev_initialize_stop(vd, VDEV_INITIALIZE_SUSPENDED, vd_list);
+		break;
+	default:
+		panic("invalid cmd_type %llu", (unsigned long long)cmd_type);
+	}
+	mutex_exit(&vd->vdev_initialize_lock);
+
+	return (0);
+}
+
+int
+spa_vdev_initialize(spa_t *spa, nvlist_t *nv, uint64_t cmd_type,
+    nvlist_t *vdev_errlist)
+{
+	int total_errors = 0;
+	list_t vd_list;
+
+	list_create(&vd_list, sizeof (vdev_t),
+	    offsetof(vdev_t, vdev_initialize_node));
+
+	/*
+	 * We hold the namespace lock through the whole function
+	 * to prevent any changes to the pool while we're starting or
+	 * stopping initialization. The config and state locks are held so that
+	 * we can properly assess the vdev state before we commit to
+	 * the initializing operation.
+	 */
+	mutex_enter(&spa_namespace_lock);
+
+	for (nvpair_t *pair = nvlist_next_nvpair(nv, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(nv, pair)) {
+		uint64_t vdev_guid = fnvpair_value_uint64(pair);
```

Comment:
This `fnvpair_value_uint64()` call can be used to panic the kernel from userland (though only root should be able to do it):

```
root@linux:/usr/src/zfs# cat /proc/sys/kernel/spl/gitrev
zfs-0.8.0-rc2-46-g9e4a98f
root@linux:/usr/src/zfs# git diff
diff --git a/lib/libzfs/libzfs_pool.c b/lib/libzfs/libzfs_pool.c
index f799471..a48cf0f 100644
--- a/lib/libzfs/libzfs_pool.c
+++ b/lib/libzfs/libzfs_pool.c
@@ -2146,7 +2146,7 @@ zpool_initialize(zpool_handle_t *zhp, pool_initialize_func_t cmd_type,
                }
 
                uint64_t guid = fnvlist_lookup_uint64(tgt, ZPOOL_CONFIG_GUID);
-               fnvlist_add_uint64(vdev_guids, vd_path, guid);
+               fnvlist_add_string(vdev_guids, vd_path, vd_path);
 
                (void) snprintf(msg, sizeof (msg), "%llu", (u_longlong_t)guid);
                fnvlist_add_string(guids_to_paths, msg, vd_path);
root@linux:/usr/src/zfs# zpool initialize testpool &
[1] 616
root@linux:/usr/src/zfs# 
Message from syslogd@linux at Jan  3 20:59:34 ...
 kernel:[   97.189554] VERIFY3(0 == nvpair_value_uint64(nvp, &rv)) failed (0 == 22)

Message from syslogd@linux at Jan  3 20:59:34 ...
 kernel:[   97.192212] PANIC at fnvpair.c:570:fnvpair_value_uint64()

root@linux:/usr/src/zfs# dmesg
[   67.876003] zavl: module license 'CDDL' taints kernel.
[   67.876008] Disabling lock debugging due to kernel taint
[   68.848431] [sched_delayed] sched: RT throttling activated
[   70.876414] ZFS: Loaded module v0.8.0-rc2_46_g9e4a98f (DEBUG mode), ZFS pool version 5000, ZFS filesystem version 5
[   83.079393] random: nonblocking pool is initialized
[   97.189554] VERIFY3(0 == nvpair_value_uint64(nvp, &rv)) failed (0 == 22)
[   97.192212] PANIC at fnvpair.c:570:fnvpair_value_uint64()
[   97.194322] Showing stack for process 616
[   97.194329] CPU: 0 PID: 616 Comm: zpool Tainted: P           O  3.16.0-4-amd64 #1 Debian 3.16.51-3
[   97.194333] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[   97.194335]  0000000000000000 ffffffff8151ea00 ffff880036e37700 ffff88003a707a80
[   97.194340]  ffffffffa0402421 ffff88003b6ab5d0 ffff88003a707c18 ffffffffa0402535
[   97.194343]  ffffffffa04a9e28 0000023a30bb2283 ffffffffa04aa6b0 ffffffffa04a8b18
[   97.194346] Call Trace:
[   97.194353]  [<ffffffff8151ea00>] ? dump_stack+0x5d/0x78
[   97.194363]  [<ffffffffa0402421>] ? spl_dumpstack+0x3c/0x3e [spl]
[   97.194369]  [<ffffffffa0402535>] ? spl_panic+0x112/0x14b [spl]
[   97.194375]  [<ffffffffa049f297>] ? nvt_add_nvpair+0xd5/0x1f4 [znvpair]
[   97.194382]  [<ffffffffa049f03a>] ? nvt_remove_nvpair+0x72/0x1fa [znvpair]
[   97.194389]  [<ffffffffa049eae2>] ? nvt_lookup_name_type+0x1be/0x2a8 [znvpair]
[   97.194398]  [<ffffffffa04a0ff1>] ? nvpair_value_common+0x2f/0x1b1 [znvpair]
[   97.194407]  [<ffffffffa04a228b>] ? nvpair_value_uint64+0x32/0x34 [znvpair]
[   97.194415]  [<ffffffffa04a65b0>] ? fnvpair_value_uint64+0x67/0x6d [znvpair]
[   97.194504]  [<ffffffffa067de92>] ? spa_vdev_initialize+0x1d7/0x329 [zfs]
[   97.194514]  [<ffffffffa04134a1>] ? spl_vmem_alloc+0x81/0x83 [spl]
[   97.194519]  [<ffffffffa04a66c2>] ? nv_alloc_sleep_spl+0x32/0x34 [znvpair]
[   97.194524]  [<ffffffffa049e444>] ? nv_mem_zalloc+0x5c/0x62 [znvpair]
[   97.194528]  [<ffffffffa049f53b>] ? nvlist_xalloc+0xa6/0xad [znvpair]
[   97.194533]  [<ffffffffa049f493>] ? nvlist_alloc+0x35/0x37 [znvpair]
[   97.194539]  [<ffffffffa04a4686>] ? fnvlist_alloc+0x2b/0x6c [znvpair]
[   97.194596]  [<ffffffffa071ed43>] ? zfs_ioc_pool_initialize+0x163/0x1c1 [zfs]
[   97.194645]  [<ffffffffa07268d3>] ? zfsdev_ioctl+0x4f5/0x799 [zfs]
[   97.194650]  [<ffffffff811c1b2f>] ? do_vfs_ioctl+0x2cf/0x4b0
[   97.194657]  [<ffffffff810593c7>] ? __do_page_fault+0x177/0x410
[   97.194663]  [<ffffffff811c1d91>] ? SyS_ioctl+0x81/0xa0
[   97.194668]  [<ffffffff81526d58>] ? async_page_fault+0x28/0x30
[   97.194675]  [<ffffffff81524d0d>] ? system_call_fast_compare_end+0x10/0x15
root@linux:/usr/src/zfs# 
```

Reply:
Good catch, I'll add a patch to address this.